cylc/flow/pathutil.py

```diff
@@ -335,15 +335,33 @@ def remove_empty_parents(
             break
 
 
-def get_next_rundir_number(run_path):
-    """Return the new run number"""
-    run_n_path = os.path.expanduser(os.path.join(run_path, "runN"))
-    try:
+def get_next_rundir_number(run_path: Union[str, Path]) -> int:
+    """Return the new run number
+
+    Args:
+        run_path: Top level installed workflow dir
+        (often ``~/cylc-run/workflow``).
+
+    """
+    run_n_path = Path(os.path.expanduser(os.path.join(run_path, "runN")))
+    if run_n_path.exists() and run_n_path.is_symlink():
         old_run_path = os.readlink(run_n_path)
-        last_run_num = re.search(r'(?:run)(\d*$)', old_run_path).group(1)
-        return int(last_run_num) + 1
-    except OSError:
-        return 1
+        # Line below could in theory not return a match group, so mypy objects.
+        # This function unlikely to be called in circumstances where this will
+        # be a problem.
+        last_run_num = re.search(                        # type: ignore
+            r'(?:run)(\d*$)', old_run_path).group(1)     # type: ignore
+        last_run_num = int(last_run_num)
+    else:
+        # If the ``runN`` symlink has been removed, make a sensible guess at
+        # what the next numbered run should be.
+        last_run_num = 0
+        for rundir in Path(run_path).glob('run[0-9]*'):
+            highest_yet = int(str(rundir).split('run')[-1])
```

Comment:
I think it would make sense to re-use the logic above. Regex ought to be `run(\d+)$` (we want to match at least 1 digit)

Reply:
Agree.